presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java

```diff
@@ -244,13 +251,13 @@ private static void serializePrimitive(BlockBuilder blockBuilder, Object value,
             return;
         }
 
-        if (type instanceof BigintType) {
-            type.writeLong(blockBuilder, (Long) value);
+        if (type instanceof BigintType || type instanceof IntegerType || type instanceof SmallintType || type instanceof TinyintType) {
+            type.writeLong(blockBuilder, ((Number) value).longValue());
             return;
         }
 
-        if (type instanceof DoubleType) {
-            type.writeDouble(blockBuilder, (Double) value);
+        if (type instanceof DoubleType || type instanceof RealType) {
+            type.writeDouble(blockBuilder, ((Number) value).doubleValue());
```

Comment:
for RealType you need to decode the value (int bits to float)

Reply:
The value is coming from avro FLOAT, DOUBLE types so it would be a java Float or Double.

Reply:
Yep, I was wrong, separated the Double and RealType conversion. Added a test to verify.